# Esercizio con variabili ed espressioni

## Problem Solving: prima lo si fa a mano!

- un passo importantissimo dello sviluppo di un algoritmo è eseguire **a mano** l'elaborazione prevista.
- se non si è in grado di risolverlo a mano, dififcilmente si riuscirà a scriver il codice corretto e risolutivo

### Esempio di problema:
Bisogna posizionare lungo il muro una fila di piastrelle nere e bianche. Per ragioni estetiche l'architetto ha specificato che la prma e l'ultima piastrella devono essere **nere**. Il vostro compito è calcolare il **numero di piastrelle*+ necessarie ed il vuoto a ciascuna delle due estremità della riga, dato lo spazio disponibile e la larghezzza di ogni piastrella.

### Prima scriviamo i commmenti nel file python

- ***Dati di partenza (numerici)***:
    1. lunghezza muro, in cm (LUNGHEZZA_MURO)
    2. lato della piastrella, in cm (DIMENSIONE_PIASTRELLA) (supponiamo quadrata)
- ***Requisiti di progetto***:
    - Riempire al meglio lo spazio disponibile
    - Iniziare e terminare con una piastrella nera
    - lasciare lo stesso spazio a sinistra e destra
- ***Dati da calcolare***:
    - numero di piastrelle bianche (num_piastrelle_b)
    - numero di pisastrelle nere (numn_piastrelle_n)
    - numero di piastrelle totali
    - spazio da lasciare su ciascun lato, cm (spazio_vuoto)

### Esempio ragionamento mentale

```pytyhon
LUNGHEZZA_MURO = 310
DIM_PIASTRELLA = 25
```

|...[N]   [B]   [N]   [B]   [N]   [B]   [N]...|

Spazio occupato dalle piastrelle = (1 + 2 volte le bianche) * DIM_PIASTRELLA

spazio occupato =?= LUNGHEZZA_MURO

piastrelle totali = LUNGHEZZA_MURO / DIM_PIASTRELLA

310 / 25 = 12.4 "piastrelle"

--> 11 piastrelle totale di cui 5 bianche e 6 nere

Quale formula per passare da 12.4 a 11?

Formula risolutiva 1:

piastrelle_totali = int(LUNGHEZZA_MURO / DIM_PIASTRELLA)

se piastrelle_totale è pari, allora

piastrelle_totali = piastrelle_totali -1

num_piastrelle_b = (piastrelle_totali-1) / 2

## Scriviamo il codice Python risolutivo
> manca il caso in cui il muro sia più piccolo della piastrella

In [13]:
# Definisco i dati di ingresso
DIM_PIASTRELLA = 25
LUNGHEZZA_MURO = 310

# calcolo spazio da lasciare ai lati
# spazio_vuoto = (LUNGHEZZA_MURO - totale_piastrelle * DIM_PIASTRELLA) / 2

# calcolo numero di piastrelle bianche, nere, totali
coppie_piastrelle = (LUNGHEZZA_MURO - DIM_PIASTRELLA) // (2 * DIM_PIASTRELLA)
num_piastrelle_b = coppie_piastrelle
num_piastrelle_n = num_piastrelle_b + 1
totale_piastrelle = num_piastrelle_b + num_piastrelle_n
spazio_vuoto = (LUNGHEZZA_MURO - totale_piastrelle * DIM_PIASTRELLA) / 2


# stampo i risultati ottenuti
print("Dato un muro di", LUNGHEZZA_MURO, "cm ed una piastrella di", DIM_PIASTRELLA, "cm")
print("dovrai posizionare", totale_piastrelle, "piastrelle, di cui")
print("piastrelle bianche", num_piastrelle_b)
print("piastrelle nere", num_piastrelle_n)
print("Dovrai lasciare", spazio_vuoto, "cm da ciascun lato!")

Dato un muro di 310 cm ed una piastrella di 25 cm
dovrai posizionare 11 piastrelle, di cui
piastrelle bianche 5
piastrelle nere 6
Dovrai lasciare 17.5 cm da ciascun lato!


> Il risultato è accettabile? 
> Si perchè anche se 17.5 * 2 è maggiore di 1 piastrella, noi possiamo togliere ed aggiungere SOLO a coppie (nel nostro caso 50 cm)

Il procedimento corretto prevede:
- casi generali -> verifica del risultato
- casi limite -> verifica del risultato
- gestione casi che generano risultati NON coerenti (piastrella inferiore al muro!) 